File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        char[] chars = s.toCharArray();
        int left = 0;
        int right=s.length() -1;
        char temp;
        
        while (left<right){
            if (isVowel(chars[left]) && isVowel(chars[right])){
                temp = chars[left];
                chars[left] = chars[right];
                chars[right] = temp;
                left++;
                right--;
            }
            else if (isVowel(chars[left])){
                right--;
            }
            else if (isVowel(chars[right])){
                left++;
            }
            else{
                left++;
                right--;
            }
        }
        return new String(chars);
    }
    private boolean isVowel(char c){
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
               c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
    }
}

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU")
        str_list = list(s)
        left=0
        right=len(str_list)-1
        
        while left<right:
           
            if str_list[left] in vowels and str_list[right] in vowels:
                str_list[left], str_list[right] = str_list[right], str_list[left]
                left+=1
                right-=1
            elif str_list[left] in vowels:
                right-=1
            elif str_list[right] in vowels:
                left+=1
            else:
                left+=1
                right-=1
        return ''.join(str_list)